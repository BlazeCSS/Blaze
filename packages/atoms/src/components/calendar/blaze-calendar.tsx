import { Component, Prop, State, Event } from '@stencil/core';
import { EventEmitter } from 'events';

@Component({
  tag: 'blaze-calendar'
})
export class Calendar {
  @Prop()
  date: string;

  @State()
  _date: Date = new Date();

  @State()
  _selectedDate: Date;

  @Event()
  onSelect: EventEmitter;

  days: Array<string> = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];

  months: Array<string> = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December'
  ];

  componentWillLoad() {
    if (this.date) {
      this._date = new Date(this.date);
      this._selectedDate = this._date;
    }
  }

  getMonthName() {
    return this.months[this._date.getMonth()];
  }

  getFirstDay() {
    return new Date(this._date.getFullYear(), this._date.getMonth(), 1).getDay();
  }

  getLastDay() {
    return new Date(this._date.getFullYear(), this._date.getMonth() + 1, 0).getDay();
  }

  getTotalDaysInMonth(diff: number = 0) {
    return new Date(this._date.getFullYear(), this._date.getMonth() + 1 + diff, 0).getDate();
  }

  selectDate(date) {
    this._selectedDate = date;
    this.onSelect.emit(date);
  }

  renderDayButton(date: Date) {
    const isToday = date.toDateString() === new Date().toDateString();
    const isSelected = date.toDateString() === this._selectedDate.toDateString();

    const inMonthClass = date.getMonth() === this._date.getMonth() ? 'c-calendar__date--in-month' : '';
    const selectedClass = isSelected ? 'c-calendar__date--selected' : '';

    return (
      <button
        class={`c-calendar__date ${inMonthClass} ${selectedClass}`}
        aria-current={isToday && 'date'}
        aria-selected={isSelected.toString()}
        onClick={() => this.selectDate(date)}>
        {date.getDate()}
      </button>
    );
  }

  populateDaysPreviousMonth() {
    const days: Array<object> = [];
    const totalDaysInPreviousMonth = this.getTotalDaysInMonth(-1);
    for (let i = totalDaysInPreviousMonth; i > totalDaysInPreviousMonth - this.getFirstDay(); i--) {
      const date = new Date(this._date);
      date.setMonth(this._date.getMonth() - 1);
      date.setDate(i);
      const day = new Date(date);
      days.unshift(this.renderDayButton(day));
    }

    return days;
  }

  populateDaysCurrentMonth() {
    const days: Array<object> = [];
    const totalDaysInMonth: number = new Date(this._date.getFullYear(), this._date.getMonth() + 1, 0).getDate();
    for (let i = 1; i <= totalDaysInMonth; i++) {
      const date = new Date(this._date);
      date.setDate(i);
      const day = new Date(date);
      days.push(this.renderDayButton(day));
    }

    return days;
  }

  populateDaysNextMonth() {
    const days: Array<object> = [];
    for (let i = 1; i < 7 - this.getLastDay(); i++) {
      const date = new Date(this._date);
      date.setDate(i);
      date.setMonth(this._date.getMonth() + 1);
      const day = new Date(date);
      days.push(this.renderDayButton(day));
    }

    return days;
  }

  navYear(diff) {
    const date = new Date(this._date);
    date.setFullYear(this._date.getFullYear() + diff);
    this._date = new Date(date);
  }

  navMonth(diff) {
    const date = new Date(this._date);
    date.setMonth(this._date.getMonth() + diff);
    this._date = new Date(date);
  }

  today() {
    this.selectDate(new Date());
    this._date = new Date();
  }

  render() {
    return (
      <div class="c-calendar">
        <button class="c-calendar__control" onClick={() => this.navYear(-1)}>
          &lsaquo;
        </button>
        <div class="c-calendar__header">{this._date.getFullYear()}</div>
        <button class="c-calendar__control" onClick={() => this.navYear(1)}>
          &rsaquo;
        </button>

        <button class="c-calendar__control" onClick={() => this.navMonth(-1)}>
          &lsaquo;
        </button>
        <div class="c-calendar__header">{this.getMonthName()}</div>
        <button class="c-calendar__control" onClick={() => this.navMonth(1)}>
          &rsaquo;
        </button>

        {this.days.map((day) => (
          <div class="c-calendar__day">{day}</div>
        ))}

        {this.populateDaysPreviousMonth()}

        {this.populateDaysCurrentMonth()}

        {this.populateDaysNextMonth()}

        <button class="c-button c-button--block c-button--info" onClick={() => this.today()}>
          Today
        </button>
      </div>
    );
  }
}
