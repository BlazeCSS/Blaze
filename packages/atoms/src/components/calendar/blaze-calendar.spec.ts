import { TestWindow } from '@stencil/core/testing';
import { Calendar } from './blaze-calendar';

describe('Calendar', () => {
  it('should build', () => {
    expect(new Calendar()).toBeTruthy();
  });

  let element;

  describe('renders', () => {
    const snapIt = (name, html) => {
      it(name, async () => {
        const window = new TestWindow();
        element = await window.load({
          components: [Calendar],
          html
        });
        window.flush();

        expect(element).toMatchSnapshot();
      });
    };

    snapIt('a default calendar', '<blaze-calendar date="1982, January 14"></blaze-calendar>');
    snapIt('a coloured calendar', '<blaze-calendar date="1982, January 14"></blaze-calendar>');
  });

  it('displays the correct date provided', async () => {
    const window = new TestWindow();
    element = await window.load({
      components: [Calendar],
      html: '<blaze-calendar date="1982, January 14"></blaze-calendar>'
    });
    window.flush();

    expect(element.querySelector('.c-calendar__header--year').innerHTML).toEqual('1982');
    expect(element.querySelector('.c-calendar__header--month').innerHTML).toEqual('January');
    expect(element.querySelector('button[aria-selected="true"]').innerHTML).toEqual('14');
  });

  it('changes years', async () => {
    const window = new TestWindow();
    element = await window.load({
      components: [Calendar],
      html: '<blaze-calendar date="1982, January 14"></blaze-calendar>'
    });
    window.flush();

    element.querySelector('.c-calendar__control--prev-year').click();
    window.flush();

    expect(element.querySelector('.c-calendar__header--year').innerHTML).toEqual('1981');

    element.querySelector('.c-calendar__control--next-year').click();
    element.querySelector('.c-calendar__control--next-year').click();
    window.flush();

    expect(element.querySelector('.c-calendar__header--year').innerHTML).toEqual('1983');
  });

  it('changes months', async () => {
    const window = new TestWindow();
    element = await window.load({
      components: [Calendar],
      html: '<blaze-calendar date="1982, January 14"></blaze-calendar>'
    });
    window.flush();

    element.querySelector('.c-calendar__control--prev-month').click();
    window.flush();

    expect(element.querySelector('.c-calendar__header--month').innerHTML).toEqual('December');

    element.querySelector('.c-calendar__control--next-month').click();
    element.querySelector('.c-calendar__control--next-month').click();
    window.flush();

    expect(element.querySelector('.c-calendar__header--month').innerHTML).toEqual('February');
  });

  it('triggers onSelect event', async (done) => {
    const window = new TestWindow();
    element = await window.load({
      components: [Calendar],
      html: '<blaze-calendar date="1982, January 14"></blaze-calendar>'
    });
    window.flush();

    element.addEventListener('onSelect', (e) => {
      try {
        expect(e.detail.toDateString()).toBe(new Date('1982, January 1').toDateString());
        done();
      } catch (err) {
        done.fail(err);
      }
    });

    element.querySelector('button.c-calendar__date.c-calendar__date--in-month').click();
  });

  it('today button changed date', async (done) => {
    const window = new TestWindow();
    element = await window.load({
      components: [Calendar],
      html: '<blaze-calendar date="1982, January 14"></blaze-calendar>'
    });
    window.flush();

    element.addEventListener('onSelect', (e) => {
      try {
        expect(e.detail.toDateString()).toBe(new Date().toDateString());
        done();
      } catch (err) {
        done.fail(err);
      }
    });

    element.querySelector('.c-calendar__today').click();
  });
});
