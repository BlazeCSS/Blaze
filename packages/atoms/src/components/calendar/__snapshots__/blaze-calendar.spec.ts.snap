// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar renders a coloured calendar 1`] = `
<blaze-calendar
  class="hydrated"
  date="1982, January 14"
  type="info"
>
  
  <div
    class="c-calendar"
  >
    <button
      class="c-calendar__control c-calendar__control--prev-year"
    >
      ‹
    </button>
    <div
      class="c-calendar__header c-calendar__header--year"
    >
      1982
    </div>
    <button
      class="c-calendar__control c-calendar__control--next-year"
    >
      ›
    </button>
    <button
      class="c-calendar__control c-calendar__control--prev-month"
    >
      ‹
    </button>
    <div
      class="c-calendar__header c-calendar__header--month"
    >
      January
    </div>
    <button
      class="c-calendar__control c-calendar__control--next-month"
    >
      ›
    </button>
    <div
      class="c-calendar__day"
    >
      Su
    </div>
    <div
      class="c-calendar__day"
    >
      Mo
    </div>
    <div
      class="c-calendar__day"
    >
      Tu
    </div>
    <div
      class="c-calendar__day"
    >
      We
    </div>
    <div
      class="c-calendar__day"
    >
      Th
    </div>
    <div
      class="c-calendar__day"
    >
      Fr
    </div>
    <div
      class="c-calendar__day"
    >
      Sa
    </div>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      27
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      28
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      29
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      30
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      31
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      1
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      2
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      3
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      4
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      5
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      6
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      7
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      8
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      9
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      10
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      11
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      12
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      13
    </button>
    <button
      aria-selected="1"
      class="c-calendar__date c-calendar__date--in-month c-button c-button--info"
    >
      14
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      15
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      16
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      17
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      18
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      19
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      20
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      21
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      22
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      23
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      24
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      25
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      26
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      27
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      28
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      29
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      30
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      31
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      1
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      2
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      3
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      4
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      5
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      6
    </button>
    <div
      class="c-calendar__footer"
    >
      <button
        class="c-calendar__today c-button c-button--block c-button--info"
      >
        Today
      </button>
    </div>
  </div>
</blaze-calendar>
`;

exports[`Calendar renders a default calendar 1`] = `
<blaze-calendar
  class="hydrated"
  date="1982, January 14"
>
  
  <div
    class="c-calendar"
  >
    <button
      class="c-calendar__control c-calendar__control--prev-year"
    >
      ‹
    </button>
    <div
      class="c-calendar__header c-calendar__header--year"
    >
      1982
    </div>
    <button
      class="c-calendar__control c-calendar__control--next-year"
    >
      ›
    </button>
    <button
      class="c-calendar__control c-calendar__control--prev-month"
    >
      ‹
    </button>
    <div
      class="c-calendar__header c-calendar__header--month"
    >
      January
    </div>
    <button
      class="c-calendar__control c-calendar__control--next-month"
    >
      ›
    </button>
    <div
      class="c-calendar__day"
    >
      Su
    </div>
    <div
      class="c-calendar__day"
    >
      Mo
    </div>
    <div
      class="c-calendar__day"
    >
      Tu
    </div>
    <div
      class="c-calendar__day"
    >
      We
    </div>
    <div
      class="c-calendar__day"
    >
      Th
    </div>
    <div
      class="c-calendar__day"
    >
      Fr
    </div>
    <div
      class="c-calendar__day"
    >
      Sa
    </div>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      27
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      28
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      29
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      30
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      31
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      1
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      2
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      3
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      4
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      5
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      6
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      7
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      8
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      9
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      10
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      11
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      12
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      13
    </button>
    <button
      aria-selected="1"
      class="c-calendar__date c-calendar__date--in-month c-button c-button--brand"
    >
      14
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      15
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      16
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      17
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      18
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      19
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      20
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      21
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      22
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      23
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      24
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      25
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      26
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      27
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      28
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      29
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      30
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date c-calendar__date--in-month"
    >
      31
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      1
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      2
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      3
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      4
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      5
    </button>
    <button
      aria-selected="0"
      class="c-calendar__date"
    >
      6
    </button>
    <div
      class="c-calendar__footer"
    >
      <button
        class="c-calendar__today c-button c-button--block c-button--brand"
      >
        Today
      </button>
    </div>
  </div>
</blaze-calendar>
`;
